pkg/stanza/operator/input/namedpipe/input.go

```diff
@@ -55,59 +58,61 @@ func (i *Input) Start(_ operator.Persister) error {
 		return fmt.Errorf("failed to chmod named pipe: %w", chmodErr)
 	}
 
-	watcher, err := NewWatcher(i.path)
-	if err != nil {
-		return fmt.Errorf("failed to create watcher: %w", err)
-	}
-
+	// Open the pipe with O_RDWR so it won't block on opening the pipe when there is no writer.
+	// The current process is both a writer and reader, which prevents the read from receiving
+	// EOF because there is always a writer (the process itself) connects to the pipe.
 	pipe, err := os.OpenFile(i.path, os.O_RDWR, os.ModeNamedPipe)
 	if err != nil {
 		return fmt.Errorf("failed to open named pipe: %w", err)
 	}
-
 	i.pipe = pipe
 
 	ctx, cancel := context.WithCancel(context.Background())
 	i.cancel = cancel
 
-	i.wg.Add(2)
-	go func() {
-		defer i.wg.Done()
-		if err := watcher.Watch(ctx); err != nil {
-			i.Logger().Error("failed to watch named pipe", zap.Error(err))
-		}
-	}()
-
-	go func() {
-		defer i.wg.Done()
-		for {
-			select {
-			case <-watcher.C:
-				if err := i.process(ctx, pipe); err != nil {
-					i.Logger().Error("failed to process named pipe", zap.Error(err))
-				}
-			case <-ctx.Done():
-				return
-			}
-		}
-	}()
+	i.wg.Add(1)
+	go i.readLoop(ctx)
 
 	return nil
 }
 
 func (i *Input) Stop() error {
-	if i.pipe != nil {
-		i.pipe.Close()
-	}
-
 	if i.cancel != nil {
 		i.cancel()
 	}
 
+	if i.pipe != nil {
+		i.pipe.Close()
+	}
+
 	i.wg.Wait()
 	return nil
 }
 
+func (i *Input) readLoop(ctx context.Context) {
+	defer i.wg.Done()
+	pipe := i.pipe
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		default:
+			err := i.process(ctx, pipe)
+			if err != nil {
+				i.Logger().Error("error processing named pipe", zap.Error(err))
+			}
+
+			// The process exits due to whatever reason, wait for ReadTimeout and try again.
+			select {
+			case <-ctx.Done():
+				return
+			case <-time.After(ReadTimeout):
```

Comment:
Might be worth a log here so we can keep track of when this happens? It probably indicates wider issues

Reply:
We logged the error from the process call, but no harm adding a log here too.